nomt/src/io/linux.rs

```diff
@@ -88,17 +90,24 @@ fn run_worker(command_rx: Receiver<IoCommand>, handle_tx: Vec<Sender<CompleteIo>
                 if pending.get(completion_event.user_data() as usize).is_none() {
                     continue;
                 }
+
                 let PendingIo { command, start } =
                     pending.remove(completion_event.user_data() as usize);
 
                 stats.note_completion(start.elapsed().as_micros() as u64);
 
-                let handle_idx = command.handle;
-                let result = if completion_event.result() == -1 {
-                    Err(std::io::Error::from_raw_os_error(completion_event.result()))
-                } else {
-                    Ok(())
+                let result = match command.kind.get_result(completion_event.result()) {
+                    IoKindResult::Ok => Ok(()),
+                    IoKindResult::Err => {
+                        Err(std::io::Error::from_raw_os_error(completion_event.result()))
+                    }
+                    IoKindResult::Retry => {
+                        retries.push_back(command);
```

Comment:
Is it true that a write can be pushed back and the end of the queue for a retry? Wouldn't you possibly change the order of writes, which might lead to the last write being an incorrect one?

Reply:
The io module does not guarantee the order in which commands are performed, this is an inherent property of io_uring.  It is up to the user to send commands, wait for completion, and properly interleave them with fsyncs

Nomt's workflow is defined within sessions, which involve reads, writes, and then fsyncs. (this is a generalization, there are multiple inner modules that adhere to a similar workflow)

Reply:
Thanks, I didn't realise there's no guaranteed order. Resolving.